### PR TITLE
Add cellrangerCall

### PR DESCRIPTION
**Pull request by jonathangriffiths**
Function added with documentation and tests.

Documentation references variables with \code{} in the description, which may be counter to your style.

Tests may need expanding, but useful ones didn't jump out at me right away.

### CODE REVIEW

File: R/cellrangerCall.R [hunk 1]
```diff
@@ -0,0 +1,23 @@
+cellrangerCall <- function(m, exp.cells = 3000, upper = 0.99, lower = 0.1)
```
Comment (LTLA) [inline R/cellrangerCall.R]: Change this to `upper.pct` and `lower.prop`.

File: R/cellrangerCall.R [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  
+  if(upper > 1 | upper < 0){
+    stop("'upper' should be a numeric value between 0 and 1")
+  }
+  
+  if(lower > 1 | lower < 0){
+    stop("'lower' should be a numeric value between 0 and 1")
+  }
+  
+  libs <- Matrix::colSums(m)
```
Comment (LTLA) [inline R/cellrangerCall.R]: You shouldn't need this `::`.

File: R/cellrangerCall.R [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  
+  top <- libs[order(libs, decreasing = TRUE)][1:exp.cells]
```
Comment (LTLA) [inline R/cellrangerCall.R]: 1. Never use `1:X`, always use `seq_len(X)`.
2. Slight increase in efficiency by doing `libs[order()[seq_len(exp.cells)]]`.

File: man/cellrangerCall.rd [hunk 4]
```diff
@@ -0,0 +1,59 @@
+\name{cellrangerCall}
+\alias{cellrangerCall}
+
+\title{Call cells from number of UMIs}
+\description{Call cells according to the number of UMIs associated wiht each barcode, like CellRanger's approach}
```
Comment (LTLA) [inline man/cellrangerCall.rd]: Typo, `wiht`.

File: man/cellrangerCall.rd [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+Columns represent barcoded droplets, rows represent cells.
+The matrix should correspond to an individual sample.
+}
+\item{exp.cells}{A numeric scalar specifying the expected number of cells in this sample.}
+\item{upper}{A numeric scalar between 0 and 1 specifying the quantile of the top \code{exp.cells} to consider for the first step of the algorithm}
+\item{lower}{A numeric scalar between 0 and 1 specifying the fraction of molecules of the \code{upper} quantile result for a barcode to be called as a cell}
+}
+
+\details{
+The \code{testEmptyDrops} function will call cells based on library size similarly to CellRanger.
```
Comment (LTLA) [inline man/cellrangerCall.rd]: WTF.

File: man/cellrangerCall.rd [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+Default arguments correspond to an exact reproduction of CellRanger's algorithm, where the number of expected cells was also left at CellRanger default value.
+
+The method considers the \code{upper} quantile top \code{exp.cells} barcodes, ordered by decreasing number of UMIs, as a threshold. Any barcodes containing more molecules than \code{lower} times this threshold is considered to be a cell, and is retained for further analysis.
+
+This method may be vulnerable to calling very well-captured background RNA, or missing real cells that were poorly captured. See \code{?emptyDrops} for an alternative approach.
```
Comment (LTLA) [inline man/cellrangerCall.rd]: use `\code{?\link{emptyDrops}}`.

File: man/cellrangerCall.rd [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+\item{exp.cells}{A numeric scalar specifying the expected number of cells in this sample.}
+\item{upper}{A numeric scalar between 0 and 1 specifying the quantile of the top \code{exp.cells} to consider for the first step of the algorithm}
+\item{lower}{A numeric scalar between 0 and 1 specifying the fraction of molecules of the \code{upper} quantile result for a barcode to be called as a cell}
+}
+
+\details{
+The \code{testEmptyDrops} function will call cells based on library size similarly to CellRanger.
+Default arguments correspond to an exact reproduction of CellRanger's algorithm, where the number of expected cells was also left at CellRanger default value.
+
+The method considers the \code{upper} quantile top \code{exp.cells} barcodes, ordered by decreasing number of UMIs, as a threshold. Any barcodes containing more molecules than \code{lower} times this threshold is considered to be a cell, and is retained for further analysis.
```
Comment (LTLA) [inline man/cellrangerCall.rd]: Some sort of typo here.

File: man/cellrangerCall.rd [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+# Identify likely cell-containing droplets. 
+called <- cellrangerCall(my.counts)
+table(called)
+
+# Get matrix of called cells
+cell.counts <- my.counts[, called]
+}
+
+\references{
+10X Genomics. 2018.
```
Comment (LTLA) [inline man/cellrangerCall.rd]: There is another citation to a 10X resource somewhere in this package. Find it and use the same format.

File: tests/testthat/test-cellrangercall.R [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#Test that function runs to completion
+# library(DropletUtils); library(testthat); source("test-emptydrops.R")
+
+test_that("cellrangerCall runs to completion", {
+  #mock up counts
+  source(system.file("scripts", "mock_empty.R", package="DropletUtils"))
+  
+  out = cellrangerCall(my.counts)
```
Comment (LTLA) [inline tests/testthat/test-cellrangercall.R]: `=` this makes me sick.

**Comment by jonathangriffiths** [comment]
New version up

**Pull request merged**